# Transcript Quantifier Assessment
Assessing StringTie, Salmon, and Kallisto with simulated RNA-Seq reads with noises introduced in read- and annotation-level.

## Data
We used the simulated data from the transcriptional noise work of Varabyou et al **[1]**. In total, there are three tissues, and each of them has 10 samples.
Three different transcriptional noises are simulated which are **(1)** intronic, **(2)** intergenic, and **(3)** splicing noises.

![Picture 1](https://user-images.githubusercontent.com/26069376/166164388-726ead1b-cdf1-4f69-a714-ed71e2452e54.png)

---

## Experiment Design
![Picture 2](https://user-images.githubusercontent.com/26069376/166164398-3d7179b0-455e-4e28-b97e-f761e1ac168d.png)


---

## Teamwork

| Kuan-Hao Chao  | Peter Ge |
| ------------- | ------------- |
| 1. Experiment design  | 1.	Result visualization |
| 2.	Running StringTie, kallisto, and salmon  | Github Maintaining |
| 3. Github Maintaining  ||

 
---

## Reference
1. Varabyou, A., Salzberg, S. L., & Pertea, M. (2021). Effects of transcriptional noise on estimates of gene and transcript expression in RNA sequencing experiments. Genome research, 31(2), 301-308.
